File: arrays/house_robber/go/bruteforce.go
package main

type seq struct {
	index int
	sum   int
}

func robBruteforce(nums []int) int {
	sequences := genAllSequences(nums)
	max := 0

	for _, sequence := range sequences {
		if sequence[len(sequence)-1].sum > max {
			max = sequence[len(sequence)-1].sum
		}
	}

	return max
}

func genAllSequences(nums []int) [][]seq {
	switch len(nums) {
	case 0:
		return [][]seq{
			{
				{sum: 0},
			},
		}
	case 1:
		return [][]seq{
			{
				{sum: nums[0]},
			},
		}
	case 2:
		max := nums[0]
		if nums[1] > max {
			max = nums[1]
		}

		return [][]seq{
			{
				{sum: max},
			},
		}
	}

	res := [][]seq{
		{
			{
				index: 0,
				sum:   nums[0],
			},
		},
		{
			{
				index: 1,
				sum:   nums[1],
			},
		},
		{
			{
				index: 0,
				sum:   nums[0],
			},
			{
				index: 2,
				sum:   nums[2] + nums[0],
			},
		},
	}

	for i := 3; i < len(nums); i++ {
		temp := make([][]seq, 0, len(res))
		for _, sequence := range res {
			if sequence[len(sequence)-1].index == i-2 || sequence[len(sequence)-1].index == i-3 {
				tmpSeq := make([]seq, 0, len(sequence))
				tmpSeq = append(sequence, seq{
					index: i,
					sum:   sequence[len(sequence)-1].sum + nums[i],
				})
				temp = append(temp, tmpSeq)
			} else {
				temp = append(temp, sequence)
			}

		}

		res = temp
	}

	return res
}

File: tasks/arrays/house_robber/ts/bruteforce.ts
type Seq = {
	index: number
	sum:   number
}

export function robBruteforce(nums: number[]): number {
	const sequences = genAllSequences(nums)
	let max = 0

	debugger;

	sequences.forEach((sequence) => {
		const lastIndex = sequence.length - 1
		if (sequence[lastIndex].sum > max) {
			max = sequence[lastIndex].sum
		}
	})

	return max
}

function genAllSequences(nums: number[]): Seq[][] {
	switch (nums.length) {
	case 0:
		return [[
			{
				index: 0,
				sum: 0,
			},
		]];
	case 1:
		return [[
			{
				index: 0,
				sum: nums[0],
			},
		]];
	case 2:
		return [[
			{
				index: 0,
				sum: Math.max(nums[0], nums[1]),
			},
		]];
	}

	let res: Seq[][] = [
		[
			{
				index: 0,
				sum:   nums[0],
			},
		],
		[
			{
				index: 1,
				sum:   nums[1],
			},
		],
		[
			{
				index: 0,
				sum:   nums[0],
			},
			{
				index: 2,
				sum:   nums[2] + nums[0],
			},
		],
	]

	for (let i = 3; i < nums.length; i++) {
		let temp: Seq[][] = []

		res.forEach((sequence) => {
			const last = sequence[sequence.length - 1]

			if (last.index === i - 2 || last.index === i - 3) {
				const tmpSeq = [
					...sequence,
					{
						index: i,
						sum:   last.sum + nums[i],
					}
				]

				temp.push(tmpSeq)
			} else {
				temp.push(sequence)
			}
		})

		res = temp
	}

	return res
}
